op-node/rollup/derive/span_batch.go

```diff
@@ -493,59 +500,88 @@ func (b *SpanBatch) GetBlockCount() int {
+	epochBit := uint(0)
+	// if there is only one batch, it is already managed by SetFirstOriginChangedBit
+	if len(b.Batches) > 1 {
+		if b.Batches[len(b.Batches)-2].EpochNum < b.Batches[len(b.Batches)-1].EpochNum {
+			epochBit = 1
+		}
+		b.originBits.SetBit(&b.originBits, len(b.Batches)-1, epochBit)
+	}
+
+	// update the blockTxCounts cache with the latest batch's tx count
+	b.blockTxCounts = append(b.blockTxCounts, uint64(len(b.Batches[len(b.Batches)-1].Transactions)))
+
+	// add each tx to the txs cache
+	// and to the spanBatchTxs cache
+	newTxs := [][]byte{}
+	for i := 0; i < len(b.Batches[len(b.Batches)-1].Transactions); i++ {
+		newTxs = append(newTxs, b.Batches[len(b.Batches)-1].Transactions[i])
+	}
+	// if the txs fail to add to the cache, store the error
+	// it can be returned when ToRawSpanBatch is called
+	// TODO: this is not ideal. If callers update caches repeatedly, they may waste time before finding out the error.
+	// in practice, ToRawSpanBatch is called after every AppendSingularBatch, so it should be fine.
+	if err := b.sbtxs.AddTxs(newTxs, b.ChainID); err != nil {
+		b.storedSbtxErr = err
+	}
 }
 
 // ToRawSpanBatch merges SingularBatch List and initialize single RawSpanBatch
-func (b *SpanBatch) ToRawSpanBatch(originChangedBit uint, genesisTimestamp uint64, chainID *big.Int) (*RawSpanBatch, error) {
+func (b *SpanBatch) ToRawSpanBatch() (*RawSpanBatch, error) {
 	if len(b.Batches) == 0 {
 		return nil, errors.New("cannot merge empty singularBatch list")
 	}
-	raw := RawSpanBatch{}
-	// Sort by timestamp of L2 block
-	sort.Slice(b.Batches, func(i, j int) bool {
-		return b.Batches[i].Timestamp < b.Batches[j].Timestamp
-	})
-	// spanBatchPrefix
+	if b.storedSbtxErr != nil {
+		return nil, b.storedSbtxErr
+	}
 	span_start := b.Batches[0]
 	span_end := b.Batches[len(b.Batches)-1]
-	raw.relTimestamp = span_start.Timestamp - genesisTimestamp
+
+	raw := RawSpanBatch{}
+	raw.relTimestamp = span_start.Timestamp - b.GenesisTimestamp
 	raw.l1OriginNum = uint64(span_end.EpochNum)
 	raw.parentCheck = b.ParentCheck
 	raw.l1OriginCheck = b.L1OriginCheck
-	// spanBatchPayload
 	raw.blockCount = uint64(len(b.Batches))
-	raw.originBits = new(big.Int)
-	raw.originBits.SetBit(raw.originBits, 0, originChangedBit)
-	for i := 1; i < len(b.Batches); i++ {
-		bit := uint(0)
-		if b.Batches[i-1].EpochNum < b.Batches[i].EpochNum {
-			bit = 1
-		}
-		raw.originBits.SetBit(raw.originBits, i, bit)
-	}
-	var blockTxCounts []uint64
-	var txs [][]byte
-	for _, batch := range b.Batches {
-		blockTxCount := uint64(len(batch.Transactions))
-		blockTxCounts = append(blockTxCounts, blockTxCount)
-		for _, rawTx := range batch.Transactions {
-			txs = append(txs, rawTx)
-		}
-	}
-	raw.blockTxCounts = blockTxCounts
-	stxs, err := newSpanBatchTxs(txs, chainID)
-	if err != nil {
-		return nil, err
-	}
-	raw.txs = stxs
+	raw.originBits = &b.originBits
+	raw.blockTxCounts = b.blockTxCounts
+	raw.txs = b.sbtxs
 	return &raw, nil
```

Comment:
I think you can convert this now to `return &RawSpanBatch{ // set fields directly }`